pepper-apis/src/main/java/org/broadinstitute/ddp/service/actvityinstanceassembler/FormInstanceCreator.java

```diff
@@ -0,0 +1,91 @@
+                instanceDto.getId(),
+                instanceDto.getActivityId(),
+                instanceDto.getActivityCode(),
+                formActivityDef.getFormType(),
+                instanceDto.getGuid(),
+                activityI18nDetail != null ? activityI18nDetail.getTitle() : null,
+                activityI18nDetail != null ? activityI18nDetail.getSubtitle() : null,
+                instanceDto.getStatusType() != null ? instanceDto.getStatusType().name() : null,
+                instanceDto.getReadonly(),
+                formActivityDef.getListStyleHint(),
+                formActivityDef.getReadonlyHintTemplate() != null ? formActivityDef.getReadonlyHintTemplate().getTemplateId() : null,
+                formActivityDef.getIntroduction() != null ? formActivityDef.getIntroduction().getSectionId() : null,
+                formActivityDef.getClosing() != null ? formActivityDef.getClosing().getSectionId() : null,
+                instanceDto.getCreatedAtMillis(),
+                instanceDto.getFirstCompletedAt(),
+                formActivityDef.getLastUpdatedTextTemplate() != null ? formActivityDef.getLastUpdatedTextTemplate().getTemplateId() : null,
+                formActivityDef.getLastUpdated(),
+                formActivityDef.canDeleteInstances(),
+                formActivityDef.isFollowup(),
+                instanceDto.isHidden(),
+                formActivityDef.isExcludeFromDisplay(),
+                instanceDto.getSectionIndex()
+        );
+        return formInstance;
+    }
+
+    private ActivityI18nDetail getActivityI18nDetail() {
+        Optional<ActivityI18nDetail> activityI18nDetail =
+                context.getHandle().attach(ActivityI18nDao.class).findDetailByLanguageCodeIdAndActivityId(
+                        context.getLangCodeId(), context.getActivityInstanceDto().getActivityId());
+        return activityI18nDetail.get();
+    }
+
+    private void addChildren(FormInstance formInstance) {
+        FormActivityDef formActivityDef = context.getFormActivityDef();
+        FormSectionCreator formSectionCreator = new FormSectionCreator(context);
+        formInstance.setIntroduction(formSectionCreator.createSection(formActivityDef.getIntroduction()));
+        formInstance.setClosing(formSectionCreator.createSection(formActivityDef.getClosing()));
+        formActivityDef.getSections().forEach(s -> {
+            formInstance.getBodySections().add(formSectionCreator.createSection(s));
+        });
+    }
+
+    private void updateBlockStatuses(FormInstance formInstance) {
+        formInstance.updateBlockStatuses(
+                context.getHandle(),
+                context.getInterpreter(),
+                context.getUserGuid(),
+                context.getOperatorGuid(),
+                context.getActivityInstanceDto().getGuid(),
+                null);
+    }
+
+    private void render(FormInstance formInstance, FormActivityDef formActivityDef) {
+        // TODO
```

Comment:
@yufengwng could you advise about rendering of FormInstance templates.
There is a method FormInstance.renderContent() where rendered `readonlyHint`, `activityDefinitionLastUpdatedText`, - what is best way to render then without excessive.
I believe it is not possible to avoid DB calls in this case. What do you think?
And another question about rendering: some of ..Def classes contains only templateID-s (and Template objects are not assigned - i.e. nulls): does it mean that it needs to query DB to get Template data? I didn't found any cache (map) with templates from where it could be fetched by IDs.